algosdk/abi/type.py

```diff
@@ -0,0 +1,1026 @@
+                else:
+                    curr_len = element.byte_len()
+                    value_partitions.append(
+                        tuple_string[array_index : array_index + curr_len]
+                    )
+                    array_index += curr_len
+            if (
+                array_index >= len(tuple_string)
+                and i != len(tuple_elements) - 1
+            ):
+                raise error.ABIEncodingError(
+                    "input string is not long enough to be decoded: {}".format(
+                        tuple_string
+                    )
+                )
+            i += 1
+
+        if len(dynamic_segments) > 0:
+            dynamic_segments[len(dynamic_segments) - 1][1] = len(tuple_string)
+            array_index = len(tuple_string)
+        if array_index < len(tuple_string):
+            raise error.ABIEncodingError(
+                "input string was not fully consumed: {}".format(tuple_string)
+            )
+
+        # Check dynamic element partitions
+        segment_index = 0
+        for i, element in enumerate(tuple_elements):
+            if element.is_dynamic():
+                value_partitions[i] = tuple_string[
+                    dynamic_segments[segment_index][0] : dynamic_segments[
+                        segment_index
+                    ][1]
+                ]
+                segment_index += 1
+
+        # Decode individual tuple elements
+        values = list()
+        for i, element in enumerate(tuple_elements):
+            val = element.decode(value_partitions[i])
+            values.append(val)
+        return values
+
+
+def type_from_string(s):
+    """
+    Convert a valid ABI string to a corresponding ABI type.
+    """
+    if s.endswith("[]"):
+        array_arg_type = type_from_string(s[:-2])
+        return ArrayDynamicType(array_arg_type)
+    elif s.endswith("]"):
+        static_array_regex = "^([a-z\d\[\](),]+)\[([1-9][\d]*)]$"
+        matches = re.search(static_array_regex, s)
+        try:
+            static_length = int(matches.group(2))
+            array_type = type_from_string(matches.group(1))
+            return ArrayStaticType(array_type, static_length)
+        except Exception as e:
+            raise error.ABITypeError(
+                "malformed static array string: {}".format(s)
+            ) from e
+    elif s.startswith("uint"):
+        try:
+            if not s[4:].isdecimal():
+                raise error.ABITypeError(
+                    "uint string does not contain a valid size: {}".format(s)
+                )
+            type_size = int(s[4:])
+            return UintType(type_size)
+        except Exception as e:
+            raise error.ABITypeError(
+                "malformed uint string: {}".format(s)
+            ) from e
+    elif s == "byte":
+        return ByteType()
+    elif s.startswith("ufixed"):
+        ufixed_regex = "^ufixed([1-9][\d]*)x([1-9][\d]*)$"
+        matches = re.search(ufixed_regex, s)
```

Comment:
nit: can move `ufixed_regex` and `static_array_regex` to global, and match symbols with compiled regex exprs. 
There's some minor benchmark for doing so in python: https://stackoverflow.com/questions/452104/is-it-worth-using-pythons-re-compile